### PR TITLE
fix micro batch training issue in DPO training

### PR DESCRIPTION
**Pull request by hmhuy0**
## Issue
When experimenting with offline DPO, I observed unexpected patterns in learning and LR scheduling:

- With the cosine LR scheduler, `min_lr` should be 10% of `learning_rate` by the end of training. However, it stays almost equal to `learning_rate`.  
- With a single GPU, the policy fails to learn.

---

## Fix for LR Scheduling
To address the scheduling issue, I made the following changes:

- Set `bs = train_batch_size // #gpus` instead of using `rollout_batch_size_per_device`.  
  - The latter doesn’t fit well in the offline learning context.  
- Set `update_interval = 1` (removed the conditional check in the offline training file).  
  - Since offline algorithms don’t have an online rollout process, `update_interval` should always be 1.  
  - A higher value causes data from the dataset to be skipped.

After these changes, the LR scheduling now behaves correctly.  
image

---

## Learning Issue with 1 GPU
Even with the LR fix, performance remains poor in the single-GPU case.  
However, switching from **Zero-2** to **Zero-3** resolves the issue when using:

```
--zero-stage 3 \
--no-use_fused_lm_head \
```

| | zero-2 | zero-3 | 
|---|---|---| 
| use_fused_lm_head=True | bad | - | 
| use_fused_lm_head=False | bad| good |
---

I am still running some ablation experiments to debug. Will update once the experiment is done.

### CODE REVIEW

**Comment by cameron-chen** [comment]
Hi Huy, thank you for bringing this out! The lr scheduling curve seems to be correct now. 

`update_interval = 1` makes sense to me in the offline setting. But I am not very sure if we should use `bs = train_batch_size // #gpus `. @lkevinzc , do you have any idea for bs?

**Comment by hmhuy0** [comment]
Hi @cameron-chen and @lkevinzc,

The figure below shows the results I got from experiments with zero-2 and zero-3 in the single GPU setting. From the args file, you guys asserted that when using zero-3, fuse_lm_head cannot be enabled, so I didn’t run experiments with it.

The results clearly show that zero-3 is able to learn, while zero-2 (with or without fuse_lm_head) is not able to learn.

image

**Comment by lkevinzc** [comment]
This is really weird to me, and tbh I haven't tested extensively on single GPU settings. I think it might take some time for me to reproduce the zero-2/3 issue (which I suspect is due to deepspeed backend because the zero-stage is the only think we modify). Therefore, I'd like to create an issue on this, and merge this PR regarding the batch size fix.

Please let us know if you have any clue on the zero-2/3 issue before I have time to investigate, many thanks! @hmhuy0
